nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,313 @@
+        self.task: Optional[asyncio.Task] = None
+
+    async def initialize(self):
+        self.storage = await get_storage(service_name=SERVICE_NAME)
+
+    async def finalize(self):
+        if self.task is not None and not self.task.done():
+            try:
+                await self.task
+            except Exception as exc:
+                logger.warning(
+                    "Indexing task raised an error while finalizing indexer",
+                    exc_info=exc,
+                )
+
+    async def index_soon(self, work: WorkUnit):
+        await self.work_queue.put(work)
+
+        if self.task is None:
+            task = asyncio.create_task(self._work_until_finish())
+            task.add_done_callback(partial(self._done_callback, self))
+            self.task = task
+
+    @property
+    def working(self) -> bool:
+        return self.task is not None
+
+    async def _work_until_finish(self):
+        while not self.work_queue.empty():
+            work = await self.work_queue.get()
+            await self._do_work(work)
+
+    def _done_callback(self, task: asyncio.Task):
+        self.task = None
+
+        # propagate errors
+        if task.exception() is not None:
+            raise task.exception()  # type: ignore
+
+    async def _do_work(self, work: WorkUnit):
+        start = time.time()
+        try:
+            logger.info(
+                f"Working on message for shard {work.index_message.shard} (seqid={work.seqid})",
+                extra={
+                    "seqid": work.seqid,
+                    "shard": work.index_message.shard,
+                    "storage_key": work.index_message.storage_key,
+                },
+            )
+            await self._index_message(work.index_message)
+        except Exception as exc:
+            event_id = capture_exception(exc)
+            logger.error(
+                "An error happened on indexer. Check sentry for more details. "
+                f"Event id: {event_id}",
+                extra={
+                    "seqid": work.seqid,
+                    "shard": work.index_message.shard,
+                    "storage_key": work.index_message.storage_key,
+                },
+                exc_info=exc,
+            )
+            await work.nats_msg.nak()
+        else:
+            await work.nats_msg.ack()
+            await signals.successful_indexing.dispatch(
+                SuccessfulIndexingPayload(
+                    seqid=work.seqid, index_message=work.index_message
+                )
+            )
+            logger.info(
+                "Message indexing finished",
+                extra={
+                    "seqid": work.seqid,
+                    "shard": work.index_message.shard,
+                    "storage_key": work.index_message.storage_key,
+                    "time": time.time() - start,
+                },
+            )
+        finally:
+            await work.mpu.end()
```

Comment:
should this be handled in a context manager in `index_message_soon`? otherwise it feels weird to start the mpu in a higher level function and ending it here.